TypeCobol/Compiler/CompilationDocumentCache.cs

```diff
@@ -49,13 +49,66 @@ public CompilationDocument GetOrAddDocument([CanBeNull] string libraryName, [Not
             return document;
         }
 
-        public void Evict([CanBeNull] string libraryName, [NotNull] string textName)
+        /// <summary>
+        /// Remove the given copies from the cache recursively: the dependent copies of the
+        /// given copies are removed too.
+        /// </summary>
+        /// <param name="textNames">List of copies to remove, each identified by its textName.</param>
+        /// <param name="evicted">Non-null set of evicted copies, to be populated by the method.</param>
+        public void RecursiveEvict([NotNull] List<string> textNames, [NotNull] HashSet<string> evicted)
         {
-            string fullName = GetFullName(libraryName, textName);
-            var keysToRemove = _documents.Keys.Where(key => key.StartsWith(fullName, StringComparison.OrdinalIgnoreCase)).ToArray();
-            foreach (var keyToRemove in keysToRemove)
+            /*
+             * TODO This method does not support qualified names for copies !
+             */
+
+            // Create a map of dependent copies: each entry gives the list of copies that include it
+            var importedBy = new Dictionary<string, HashSet<string>>(StringComparer.OrdinalIgnoreCase);
+            foreach (var document in _documents)
+            {
+                var copy = document.Value.TextSourceInfo.Name;
+                var importedCopies = document.Value.CopyTextNamesVariations;
+                foreach (var importedCopy in importedCopies)
```

Comment:
This should work right ?
Let's forget that `CopyTextNamesVariations` contains all imported copies recursively for a moment. As long as all copies in the chain are in cache, we'll go over them and build the correct reverse dependency list. So the algorithm should work.
Now, considering that `CopyTextNamesVariations` already flattens the dependencies, it means that the last part of the method is not really recurvis we'll have only one recursive call so this could be re-written in a non-recursive way.

Reply:
Right, I forgot that `CopyTextNamesVariations` flattens the dependencies.
Nonetheless, the new code is clearer about that.